core/base/src/main/java/de/mm20/launcher2/search/Location.kt

```diff
@@ -133,29 +133,21 @@ abstract class Location : SavableSearchable {
         )
     }
 
-    fun toAndroidLocation(): android.location.Location {
-        val location = android.location.Location("KvaesitsoLocationProvider")
+    fun toAndroidLocation(): AndroidLocation {
+        val location = AndroidLocation("KvaesitsoLocationProvider")
 
         location.latitude = latitude
         location.longitude = longitude
 
         return location
     }
 
-    fun distanceTo(androidLocation: android.location.Location): Float {
-        lastDistance = androidLocation.distanceTo(this.toAndroidLocation())
-        return lastDistance!!
+    fun distanceTo(androidLocation: AndroidLocation): Float {
+        return androidLocation.distanceTo(this.toAndroidLocation())
     }
 
     fun distanceTo(otherLocation: Location): Float =
         this.distanceTo(otherLocation.toAndroidLocation())
-
-    override fun compareTo(other: SavableSearchable): Int {
-        if (other !is Location || other.lastDistance == null || this.lastDistance == null)
-            return super.compareTo(other)
-
-        return lastDistance!!.compareTo(other.lastDistance!!)
-    }
```

Comment:
I suggest keeping this logic. With this override, the secondary criterion for sorting of search results is the distance; Which makes sense, since the pizza store 100m closer might be more relevant than "ABC-Pizza" that is further away. (Assuming that otherwise, entries are sorted by name alphabetically.) This ties nicely into the comparisons in `SearchVM` that defaults to `a.compareTo(b)` if neither of the searchables have a relevance.

I see that this won't work in interfaces. So maybe implement this in `OsmLocation`? I just put this into Location since distances are nothing specific to OSM, but to locations in general. (Again, assuming we would ever have another location provider :b)

Reply:
I sort of implemented a best-effort approach, by populating `lastDistance` every time `distanceTo(other)` was called:

`Location.kt`:
```kotlin
fun distanceTo(androidLocation: android.location.Location): Float {
    lastDistance = androidLocation.distanceTo(this.toAndroidLocation())
    return lastDistance!!
}
```

Now some months later, I see that this method should rather be called something like `distanceToUser()` since the logic behind it assumes that the passed location is the user's location. (This was probably implicitly given, since the parameter is of type `AndroidLocation`. But then, there is also `distanceTo(Location)`...)

Also, my approach increases that _spicy_ non-determinism. I judged `compareTo()` to be non-critical though for anything else than sorting.

Since you also seem to be in favor of caching, this seemed like the best approach to me. We compute this distance when the user is initially searching for something that is not in the database, in `OsmRepository.kt`, 140ff:

```kotlin
result
    .filter {
        !settings.hideUncategorized || (it.category != null && it.category != LocationCategory.OTHER)
    }
    .groupBy {
        it.label.lowercase()
    }
    .flatMap { (_, duplicates) ->
        // deduplicate results with same labels, if
        // - same category
        // - distance is less than 100m
        if (duplicates.size < 2) duplicates
        else {
            val luckyFirst = duplicates.first()
            duplicates
                .drop(1)
                .filter {
                    it.category != luckyFirst.category ||
                        it.distanceTo(luckyFirst) > 100.0
                } + luckyFirst
        }
    }.sortedBy {
        it.distanceTo(userLocation)
    }.toImmutableList()
```

I see that there needs to be some other approach though when we receive something that is de-serialized by the database.

Reply:
Maybe we could ask `DevicePoseProvider` for some cached location, and otherwise just don't sort them via distance? Users that have location-search / weather enabled are accessing their location all the time anyway. And in case there is no location available, we loose some UX, but nothing major breaks. 

I suppose for this to work correctly, weather-providers should also access user-pose via `DevicePoseProvider` in order to have this centralized piece of code, right? (But let's not dive into refactoring-hell amirite)

Reply:
> My approach increases that spicy non-determinism I suppose. I judged compareTo() to be non-critical though for anything else than sorting. But I see that there needs to be some other approach when we receive something that is de-serialized by the database.

Actually, this could be an issue. The [`Comparable`](https://docs.oracle.com/javase/8/docs/api/java/lang/Comparable.html) interface requires the relationship to be transitive (so that $` a < b \land b < c \implies a < c `$). This is not the case when the results of `compareTo` are not stable. When we update `lastDistance` while another thread is sorting a list of locations, the order would be inconsistent and I don't know how the Android Runtime would handle such a situation.

> Maybe we could ask DevicePoseProvider for some cached location, and otherwise just don't sort them via distance? Users that have location-search / weather enabled are accessing their location all the time anyway. And in case there is no location available, we loose some UX, but nothing major breaks.

This might be the best solution. Give the location an "assumed user location", which is constant for the lifetime of the `Location` object. Sort by that value. Realistically, the lifetime of a `Location` object isn't that long – unless it's pinned to favorites but in that case it isn't sorted by distance.

Reply:
Are only pinned/launched items saved to the launcher database? Because if not, we might have some Locations that have no sorting information, which will sort of "penalize" them for not coming from the API.

Reply:
> But in that case we can simply aquire the last known location from DevicePoseProvider and populate the restored location with that, right?

You mean during deserialization? SGTM!

Even then I'd argue we do not have a hard requirement on locations being the same, 1m difference doesn't matter. However, if for some reason the deserialization is quicker than the API query we might end up with a location for deserialized instances that is much further away than some new location returned for the API query. This does sound like an edge case though...

Reply:
> However, if for some reason the deserialization is quicker than the API query we might end up with a location for deserialized instances that is much further away than some new location returned for the API query. This does sound like an edge case though

It's pretty likely that that happens. The edge case here is that someone actually assigns a tag or custom label to a location… I would say if a location with a custom label matches the query, we just show it, no matter how far away it is. If someone goes through the steps of assigning a custom label, and then searches for a term that matches that label label, I'd imagine that they *want* to find that result. And even if not, it's just a bit more noise, it's not like it would hide any results from the API because of that.